### PR TITLE
#267 support XMP tag 'subject'

### PR DESCRIPTION
**Pull request by kagahd**
PR for supporting xmp subject as keywords.

### CODE REVIEW

File: src/backend/model/threading/MetadataLoader.ts [hunk 1]
```diff
@@ -216,7 +216,13 @@ export class MetadataLoader {
               if (exif.Rating) {
                 metadata.rating = (parseInt(exif.Rating.value, 10) as any);
               }
-
+              if(exif.subject && exif.subject.value){
+                for(let i=0; i < exif.subject.value.length; i++){
```
Comment (bpatrik) [inline src/backend/model/threading/MetadataLoader.ts:220]: add something like this here ` metadata.keywords = iptcData.keywords || [];`

File: test/backend/assets/xmp/xmp_subject.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    "fStop": 2.4,
+    "focalLength": 4.32,
+    "make": "samsung",
+    "model": "SM-G975F"
+  },
+  "creationDate": 1614703656000,
+  "fileSize": 7909,
+  "orientation": 1,
+  "rating": 50,
+  "Rating": 3,
```
Comment (bpatrik) [inline test/backend/assets/xmp/xmp_subject.json:14]: having two ratings fiels looks suspicious to me.

**Comment by bpatrik** [comment]
I added some comments.
You can try running the test locally too. Your IDE should be able to help with that (Webstorm does help for me). If not, you can try running it from terminal with something like this: `mocha test/**/*.js --grep "^MetadataLoader"` from the project root. (mocha need to be installled globally)

**Comment by bpatrik** [comment]
I will look into the errors in more details. It is is possible, that with this change the galllery will be able to read more metadata from the files, than we first expected (like reading keywords from png file which does not have exif data)
So the failing test might be features instead of bugs, but I need to look at them more closely

**Comment by kagahd** [comment]
Yes, you're right. My modification reads more metadata from the files while the tests did not reflected them. That's the case for the `test_png.png` file which has 2 keywords in the XMP `subject` section. However, I had to add `"keywords":[ ]` to the `test_png.json` because the keywords were **not** retrieved by the `ExifReader`. Do you know why? 

I found out that `metadata.keywords` is only initialized if the photo has IPTC data. If it's not, my code could not push the keywords to the (not existing) array. So I had to initialize it if not yet done. That's the reason why the keywords of my photo were not present.

The tests for the photo with special characters was failing because I wrongly used the `.value` instead of `.description` to get the value from the `subject` fields. I found out that I had to use `.description` to get right encoding. However, I still don't understand why the photo has a keyword `squirrel` in both IPTC and XMP but the json file to test against doesn't have it. Nevertheless the test passes.

To make the test pass, I removed my rating from my test photo. I found out that XnView added to the XMP meta data a value from 0 to 100 as `MicrosoftPhoto.Rating` besides the original tag `xmp.Rating` which goes only from 0 to 5. ExifReader overwrites the second last read value by the last read value. In my case, `MicrosoftPhoto.Rating` overwrote `xmp.Rating` which made the test fail.

Last but not least I did not understand why `npm test` failed locally on test `39) SearchQueryParser`, I found out that it has nothing to do with my code changes because this test fails also locally without my code changes. However, the test passes online.

Thanks @bpatrik to point me to `mocha` to to run the test locally. I didn't know about it and used locally `npm test` which takes probably much more time to execute than just to test the `MetadataLoader`. I'll try it out.

**Comment by kagahd** [comment]
Btw. `mocha test/**/*.js --grep "^MetadataLoader"` just showed `0 passing (2ms)`, so it seems that it does not run any test.

`npm test -- --grep "^MetadataLoader"` did the trick but was as slow as `npm test` (1:23 min).

**Comment by bpatrik** [comment]
Thank you for your hard work. It looks good!

>  I had to add "keywords":[ ] to the test_png.json because the keywords were not retrieved by the ExifReader. Do you know why?

I think because png does not support exif: https://stackoverflow.com/questions/9542359/does-png-contain-exif-data-like-jpg
Maybe a newer png would already.

> mocha issue

I'm not sure why it is not running for you. I usually just use my IDE to start the tests:
kép

This would run something like this on windows:
```
"<path1>\nodejs\node.exe" <path2>\PiGallery2\node_modules\mocha\bin\mocha --require ts-node/register --ui bdd --reporter <path3>\WebStorm\ch-0\211.6693.108\plugins\NodeJS\js\mocha-intellij\lib\mochaIntellijReporter.js <path2>\PiGallery2\test\backend\unit\model\threading\MetaDataLoader.spec.ts --grep "^MetadataLoader "
``` 
But running only `mocha test/**/*.js --grep "^MetadataLoader"` also worked for me. 
For issue I can imagen the working dir was not the project , you did not have mocha installed globally, the test files were not trascoded to js

**Pull request merged**